bionic/descriptors/ast.py

```diff
@@ -18,39 +18,22 @@
 """
 
 from abc import ABC, abstractmethod
+from enum import Enum
 from functools import total_ordering
 
 import attr
 
-# TODO We have a lot of code that uses isinstance() to check for different types of
-# descriptor node. I think we could avoid this by introducing a few new types of helper:
-#
-# 1. Some way to pattern-match on tuple type, like maybe:
-#
-#        value = dnode.match(
-#            entity=lambda entity_node: ...,
-#            tuple=lambda tuple_node: ...,
-#            draft=lambda draft_node: ...,
-#        )
-#
-#    I'm not sure if the above is the best way, but this would help guarantee that we
-#    handle every case. It could also look like:
-#
-#        class MyMatcher(NodeMatcher):
-#            def match_entity(self, entity_node):
-#                return ...
-#            def match_tuple(self, tuple_node):
-#                return ...
-#            ...
-#        value = MyMatcher().match(dnode)
-#
-# 2. Methods to determine if a node has a particular type:
-#
-#    dnode.is_entity()
-
-# 3. Methods to assert that a node has a particular type:
-#
-#    name = dnode.as_entity().name
+from ..utils.misc import oneline
+
+
+class Subtype(Enum):
```

Comment:
Wondering about the naming, what about just calling it `Type`? `Subtype` seems to imply there is a `Type`.

Reply:
Yeah, good question. I was reluctant to call it a `Type` because it's not literally a Python type -- I feel like it would be misleading to have a `dnode.type` field whose value isn't actually a `type` object. I agree that "subtype" is not a great name either. I guess we could call the class `DescriptorNodeType` and make the field `dnode.type_enum`. WDYT?